src/transport.rs

```diff
@@ -0,0 +1,108 @@
+    ///
+    /// The response is returned as a `Self::Stream` - some type implementing the `Read` trait. The
+    /// library will read all of the data and parse it as a response. It must be UTF-8 encoded XML,
+    /// otherwise the call will fail.
+    ///
+    /// If a transport error occurs, this should return it as a boxed error - the library will
+    /// interpret it as a transport error and return an appropriate `RequestError`.
+    fn transmit(self, request: &Request) -> Result<Self::Stream, Box<Error>>;
+}
+
+/// Contains `Transport` implementations for types of the `reqwest` crate.
+#[cfg(feature = "reqwest")]
+mod reqwest {
+    extern crate reqwest;
+
+    use {Request, Transport};
+    use self::reqwest::{Client, RequestBuilder, Url, mime};
+    use self::reqwest::header::{ContentType, ContentLength, UserAgent};
+
+    use std::error::Error;
+
+    /// Use a `RequestBuilder` as the transport.
+    ///
+    /// The request will be sent as specified in the XML-RPC specification: A default `User-Agent`
+    /// will be set, along with the correct `Content-Type` and `Content-Length`.
+    impl Transport for RequestBuilder {
+        type Stream = reqwest::Response;
+
+        fn transmit(mut self, request: &Request) -> Result<Self::Stream, Box<Error>> {
+            // First, build the body XML
+            let mut body = Vec::new();
+            // This unwrap never panics as we are using `Vec<u8>` as a `Write` implementor,
+            // and not doing anything else that could return an `Err` in `write_as_xml()`.
+            request.write_as_xml(&mut body).unwrap();
+
+            // Set all required request headers
+            // NB: The `Host` header is also required, but reqwest adds it automatically, since
+            // HTTP/1.1 requires it.
+            let builder = self
+                .header(UserAgent::new("Rust xmlrpc"))
+                .header(ContentType("text/xml; charset=utf-8".parse().unwrap()))
+                .header(ContentLength(body.len() as u64));
+
+            let response = builder
+                .body(body)
+                .send()?
+                .error_for_status()?;
+
+            // Check response headers
+            // "The Content-Type is text/xml. Content-Length must be present and correct."
+            if let Some(content) = response.headers().get::<ContentType>() {
+                // (we ignore this if the header is missing completely)
+                match (content.type_(), content.subtype()) {
+                    (mime::TEXT, mime::XML) => {},
+                    (ty, sub) => return Err(
+                        format!("expected Content-Type 'text/xml', got '{}/{}'", ty, sub).into()
+                    ),
+                }
+            }
+
+            response.headers().get::<ContentLength>()
+                .ok_or_else(|| format!("expected Content-Length header, but none was found"))?;
+
+            Ok(response)
```

Comment:
All of this can probably exposed as a helper, so users who only want to access headers (in order to use cookies or configure the user agent) can do so without having to write a large impl like this.